Add documentation - Develop On Kuzzle - Embedded SDK (#1859)
kuzzleio/kuzzle: doc/2/guides/develop-on-kuzzle/1-embedded-sdk/index.md

---
code: false
type: page
title: Embedded SDK
description: Execute API action from backend code
order: 200
---

# Embedded SDK

::: info

> The Embedded SDK is available only in the `runtime` phase, after application startup.

Reviewer: ```suggestion
The Embedded SDK is available only during the `runtime` phase, after the application has started.
```

```js
// After application startup

// [...]

// Creates a document
await app.sdk.document.create('nyc-open-data', 'yellow-taxi', {
  name: 'Aschen',
  age: 27
})
```

## Query method

<!-- Duplicate /core/2/guides/getting-started/6-write-application -->

The low level [query](/sdk/js/7/core-classes/kuzzle/query) method can also be used to **send custom requests to the Kuzzle API**.  

**Example:** _Execute a custom controller action with the [query](/sdk/js/7/core-classes/kuzzle/query) method_
```js
// After application startup

// [...]

// Execute a custom controller action
await app.sdk.query({
  controller: 'greeting',
  action: 'name',
  name: 'Aschen'
})
```

## User impersonation

By default, when using the Embedded SDK, requests **don't have the same context as the original request** received by Kuzzle.

Typically, the `request.context.user` property is not set and thus **[Kuzzle metadata](/core/2/some-link) will not be set when creating or updating documents**.

It is possible to use the same user context as the original request with the Embedded SDK, for this purpose it is necessary to use the [EmbeddedSDK.as](/core/2/some-link) method.

> **Example:** _Creating a document as the original API user to preserv Kuzzle metadata_

Reviewer: ```suggestion
**Example:** _Creating a document as the original API user to preserve Kuzzle metadata_
```

```js
app.controller.register('drivers', {
  actions: {
    create: {
      handler: async request => {
        const originalUser = request.context.user

        return app.sdk.as(originalUser).document.create(
          'nyc-open-data',
          'yellow-taxi',
          { name: 'Aschen' })
       }
    }
  }
})
```

::: warning

> User permissions are not applied even when the [EmbeddedSDK.as](/core/2/some-link) method is used.

Reviewer: I think a lengthier explanation is required here.

```suggestion
User permissions are applied only once, when a request is received by Kuzzle through the exposed API.  
If a request is authorized, then all subsequent calls to the API performed with [EmbeddedSDK.as](/core/2/some-link) are always authorized, even if they are made to execute API actions that a user is normally forbidden from.
```

:::

## Backend Realtime Subscriptions

> Realtime subscriptions should be made using the [Realtime Controller](/sdk/js/7/controllers/realtime) **just after the application startup**.

Reviewer: An explanation of the differences with external calls to `realtime:subscribe` is in order here, I felt myself a bit lost with this feature (I'm still not accustomed to it :grin:).

This explanation also helps when introducting the `propagate` option below, as it gives a bit more context to the reader.

```suggestion
Realtime subscriptions should be made using the [Realtime Controller](/sdk/js/7/controllers/realtime) **just after the application has started**.  

Realtime subscriptions performed by an application are used so that an application gets notified about changes, and can act upon them. The behavior is much the same than when a client subscribes, but since the entity performing the subscription is different (client vs. application), the feature accessible by an application has some new options to fine tune how notifications are propagated across a Kuzzle cluster.
```

Author: Thanks :D

::: warning

> You should **avoid making dynamic subscriptions at runtime because** that can lead to unwanted behavior, since the subscriptions won't be replicated on other cluster nodes.

Reviewer: ```suggestion
You should **avoid making dynamic subscriptions at runtime**, because that can lead to unwanted behavior, since those subscriptions won't be replicated to other cluster nodes.
```

:::

> The `propagate` option defines if, for that subscription, notifications should be propagated to (and processed by) all cluster nodes, or if only the node having received the triggering event should handle it.

Author: I don't really like `property` because here `propagate` is an option of the `subscribe` method (but yes a property of the option object)

Also it's still unclear because notifications are always propagated but it's just the callback who is executed either on each node or on one node

> The `propagate` option defines if, for that subscription, notifications should be propagated to (and processed by) all cluster nodes, or if only the node having received the triggering event should handle it.

Reviewer 2: I understand what you mean, it's just that after reading it few time, this sentence is hard to get

> The `propagate` option defines if, for that subscription, notifications should be propagated to (and processed by) all cluster nodes, or if only the node having received the triggering event should handle it.

Author: The `propagate` option
@scottinet Some thought about this?

### propagate: false (default)

> With `propagate: false`, the callback function is **executed only one node**, the one on which the notification is generated (only one execution).

Reviewer: ```suggestion
With `propagate: false`, the callback function is **executed only on one node**, the one on which the notification is generated (only one execution).
```

::: info 

> This behavior is suitable for most usage like sending emails, write in the database, call an external API, etc.

Reviewer: ```suggestion
This behavior is suitable for most usages such as sending emails, write in the database, call an external API, etc.
```

:::

**Example:**

```js
app.start()
  .then(async () => {
    // The default value for the "propagate" option is "false"
    await app.sdk.realtime.subscribe(
      'nyc-open-data',
      'yellow-taxi',
      {},
      notification => {
        // This callback will be executed only on the node generating a notification
      })
  })
```

### propagate: true

> With `propagate: true`, notifications are propagated to all nodes of a cluster, **executing the callback function on each nodes**.

Reviewer: ```suggestion
With `propagate: true`, notifications are propagated to all nodes of a cluster, **executing the callback function on all nodes**.
```

```js
// After application startup

// [...]

// Execute a custom controller action
await app.sdk.query({
  controller: 'greeting',
  action: 'sayHello',
  name: 'Aschen'
})
```

## User impersonation

By default, when using the Embedded SDK, requests **don't have the same context as the original request** received by Kuzzle.

Typically, the `request.context.user` property is not set and thus **[Kuzzle metadata](/core/2/some-link) will not be set when creating or updating documents**.

It is possible to use the same user context as the original request with the Embedded SDK, for this purpose it is necessary to use the [EmbeddedSDK.as](/core/2/some-link) method.

**Example:** _Creating a document as the original API user to preserve Kuzzle metadata_
```js
app.controller.register('drivers', {
  actions: {
    create: {
      handler: async request => {
        const originalUser = request.context.user

        return app.sdk.as(originalUser).document.create(
          'nyc-open-data',
          'yellow-taxi',
          { name: 'Aschen' })
       }
    }
  }
})
```

::: warning
User permissions are applied only once, when a request is received by Kuzzle through the exposed API.  

> If a request is authorized, then all subsequent calls to the API performed with [EmbeddedSDK.as](/core/2/some-link) are always authorized, even if they are made to execute API actions that a user is normally forbidden from.

Reviewer 3: This behavior is REEEEEEELY weird, no?

Reviewer: Yes. We already mentioned this issue in our product workshops, this method should instead work as a user impersonation...
But for now this is how it works.

:::

## Backend Realtime Subscriptions

Realtime subscriptions should be made using the [Realtime Controller](/sdk/js/7/controllers/realtime) **just after the application has started**.  

> Realtime subscriptions performed by an application are used so that an application gets notified about changes, and can act upon them. The behavior is much the same than when a client subscribes, but since the entity performing the subscription is different (client vs. application), the feature accessible by an application has some new options to fine tune how notifications are propagated across a Kuzzle cluster.

Reviewer 3: ```suggestion
Realtime subscriptions performed by an application are used so that an application gets notified about changes, and can act upon them. The behavior is the same as when a client subscribes, but since the entity performing the subscription is different (client vs. application), the feature accessible by an application has some new options to fine tune how notifications are propagated across a Kuzzle cluster.
```

::: warning
You should **avoid making dynamic subscriptions at runtime**, because that can lead to unwanted behavior, since those subscriptions won't be replicated to other cluster nodes.
:::

> The `propagate` option defines if, for that subscription, the callback execution should be propagated to all cluster nodes, or if only the node generating the notification should execute it's callback.

Reviewer 3: ```suggestion
The `propagate` option defines if, for that subscription, the callback execution should be propagated to all cluster nodes, or if only the node generating the notification should execute its callback.
```

### propagate: false (default)

With `propagate: false` only the node who generates the notification will execute the callback function

::: info 

> This behavior is suitable for most usage like sending emails, write to the database, call an external API, etc.

Reviewer 3: ```suggestion
This behavior is suitable for most usages like sending emails, write to the database, call an external API, etc.
```
